src/OrleansRuntime/Catalog/GrainCreator.cs

```diff
@@ -47,56 +38,13 @@ public Grain CreateGrainInstance(IGrainActivationContext context)
             grain.Runtime = this.grainRuntime.Value;
             grain.Identity = context.GrainIdentity;
 
-            return grain;
-        }
-
-        /// <summary>
-        /// Create a new instance of a grain
-        /// </summary>
-        /// <param name="context">The <see cref="IGrainActivationContext"/> for the executing action.</param>
-        /// <param name="stateType">If the grain is a stateful grain, the type of the state it persists.</param>
-        /// <param name="storage">If the grain is a stateful grain, the storage used to persist the state.</param>
-        /// <returns></returns>
-        public Grain CreateGrainInstance(IGrainActivationContext context, Type stateType, IStorage storage)
-		{
-            //Create a new instance of the grain
-            var grain = CreateGrainInstance(context);
-
-            var statefulGrain = grain as IStatefulGrain;
-
-            if (statefulGrain == null)
-                return grain;
-
-            //Inject state and storage data into the grain
-            statefulGrain.GrainState.State = Activator.CreateInstance(stateType);
-            statefulGrain.SetStorage(storage);
+            // wire up to lifecycle
+            var participant = grain as ILifecycleParticipant<GrainLifecycleStage>;
+            participant?.Participate(context.ObservableLifecycle);
```

Comment:
Moving state management logic out of catalog and grain-creator into grain base classes that participate in the lifecycle is the bulk of this change.  While we should be moving away from base classes, this preserves the existing behaviors and moves this logic out of the catalog/grain-creator and provides a more extensible pattern for future features.